# RetailSense : Explainable Multi-Horizon Retail Demand Forecasting
## Notebook 02 · Data Loading & Structural Inspection

> **Organization:** Celebal Technologies &nbsp;|&nbsp; **Intern:** Ayush Choudhary  
> **Domain:** Data Science | Time Series Forecasting | ML | Explainable AI  
> **Dataset:** Kaggle Store-Item Demand Forecasting (913,000 records · 10 Stores · 50 Items · 5 Years)  

---

### 1. Objective
Load the Kaggle Store-Item Demand dataset and perform an initial inspection of:
- File availability & path configuration
- Dataset dimensions, column schema, and data types
- Memory footprint with and without dtype optimization
- Date coverage and series composition overview

No preprocessing or feature engineering is applied in this notebook. Raw data is loaded as-is.

---


# RetailSense : Explainable Multi-Horizon Retail Demand Forecasting
## Notebook 02 - Data Loading & Structural Inspection

> **Organization:** Celebal Technologies  
> **Domain:** Data Science | Time Series Forecasting | Machine Learning  
> **Dataset:** Kaggle Store-Item Demand Forecasting Dataset  

---

### 1. Objective & Scope
The goal of this notebook is to load the Kaggle Store-Item Demand Forecasting dataset, inspect schema integrity, date continuity, column metadata, and apply automated memory downcasting to ensure computational scalability.

Key questions addressed:
- Are all training (`train.csv`) and testing (`test.csv`) files present?
- What are the dimensions, dtypes, and memory footprints of the raw datasets?
- How much memory optimization is achieved by downcasting numerical types?



In [1]:
import sys
from pathlib import Path
sys.path.append('..')

import pandas as pd
import numpy as np
from src.data.loader import load_raw_data
from src.utils.config import TRAIN_PATH, TEST_PATH

print("Loading Kaggle Store-Item Demand datasets with memory optimization...")
train_df, test_df = load_raw_data(TRAIN_PATH, TEST_PATH)

print(f"\nTraining Matrix Shape : {train_df.shape[0]:,} rows x {train_df.shape[1]} columns")
print(f"Test Matrix Shape     : {test_df.shape[0]:,} rows x {test_df.shape[1]} columns")

Loading Kaggle Store-Item Demand datasets with memory optimization...
Memory usage optimized from 27.86 MB to 10.45 MB (62.5% reduction).
Memory usage optimized from 1.37 MB to 0.60 MB (56.2% reduction).

Training Matrix Shape : 913,000 rows x 4 columns
Test Matrix Shape     : 45,000 rows x 4 columns


### 2. Dataset Previews & Field Descriptions

In [2]:
print("--- Training Set Preview ---")
display(train_df.head(10))

print("\n--- Test Set Preview ---")
display(test_df.head(10))

--- Training Set Preview ---


,date,store,item,sales
0,2013-01-01,1,1,13
1,2013-01-02,1,1,11
2,2013-01-03,1,1,14
3,2013-01-04,1,1,13
4,2013-01-05,1,1,10
5,2013-01-06,1,1,12
6,2013-01-07,1,1,10
7,2013-01-08,1,1,9
8,2013-01-09,1,1,12
9,2013-01-10,1,1,9



--- Test Set Preview ---


,id,date,store,item
0,0,2018-01-01,1,1
1,1,2018-01-02,1,1
2,2,2018-01-03,1,1
3,3,2018-01-04,1,1
4,4,2018-01-05,1,1
5,5,2018-01-06,1,1
6,6,2018-01-07,1,1
7,7,2018-01-08,1,1
8,8,2018-01-09,1,1
9,9,2018-01-10,1,1


### Field Descriptions:
- **date**: Observation timestamp (`YYYY-MM-DD`).
- **store**: Unique numerical identifier for the store (1 to 10).
- **item**: Unique numerical identifier for the product item (1 to 50).
- **sales**: Target variable representing daily unit sales count.


In [3]:
print("--- Data Types & Summary Statistics ---")
print(train_df.info())

print("\n--- Sales Statistics Breakdown ---")
display(train_df['sales'].describe().to_frame())

--- Data Types & Summary Statistics ---
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 913000 entries, 0 to 912999
Data columns (total 4 columns):
 #   Column  Non-Null Count   Dtype         
---  ------  --------------   -----         
 0   date    913000 non-null  datetime64[ns]
 1   store   913000 non-null  int8          
 2   item    913000 non-null  int8          
 3   sales   913000 non-null  int16         
dtypes: datetime64[ns](1), int16(1), int8(2)
memory usage: 10.4 MB
None

--- Sales Statistics Breakdown ---


,sales
count,913000.000000
mean,52.250287
std,28.801144
min,0.000000
25%,30.000000
50%,47.000000
75%,70.000000
max,231.000000


### 3. Date Span & Structural Summary
Notebook 02 successfully verifies that all 913,000 observations are cleanly loaded and optimized in memory. Next, Notebook 03 will conduct a comprehensive Data Quality Audit.

---
### Key Findings

| Attribute | Training Set | Test Set |
|---|---|---|
| Total Records | 913,000 | 45,000 |
| Columns | 4 (date, store, item, sales) | 4 (id, date, store, item) |
| Date Range | 2013-01-01 → 2017-12-31 | 2018-01-01 → 2018-03-31 |
| Unique Stores | 10 | 10 |
| Unique Items | 50 | 50 |
| Raw Memory | 27.9 MB | 1.4 MB |
| **Optimized Memory** | **10.5 MB (−62.5%)** | **0.6 MB (−56%)** |

- ✅ All dataset files present and correctly loaded
- ✅ Date column parsed as `datetime64` — no manual conversion needed
- ✅ Memory downcasting reduces RAM footprint by **62.5%** (crucial for large-scale pipelines)
- ✅ Dataset structure is complete and suitable for time-series forecasting

